common/src/api/external/mod.rs

```diff
@@ -802,15 +783,47 @@ impl Generation {
         // exceeds the value allowed by an i64.  But it seems unlikely enough to
         // happen in practice that we can probably feel safe with this.
         let next_gen = self.0 + 1;
-        // `as` is a little distasteful because it allows lossy conversion, but
-        // (a) we know converting `i64::MAX` to `u64` will always succeed
-        // losslessly, and (b) it allows to make this function `const`, unlike
-        // if we were to use `u64::try_from(i64::MAX).unwrap()`.
-        assert!(next_gen <= i64::MAX as u64);
+        assert!(
+            next_gen <= Generation::MAX.0,
+            "attempt to overflow generation number"
+        );
         Generation(next_gen)
     }
 }
 
+impl<'de> Deserialize<'de> for Generation {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = u64::deserialize(deserializer)?;
+        Generation::try_from(value).map_err(|GenerationOverflowError(_)| {
+            serde::de::Error::invalid_value(
+                serde::de::Unexpected::Unsigned(value),
+                &"an integer between 0 and 9223372036854775807",
```

Comment:
part of me wants to suggest that we just use `Generation::MAX` here, but I guess serde expects an `&str`?

Reply:
yep :(